Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/opensaml/CertificateUtil.java

```diff
@@ -76,4 +98,137 @@ private void closeStream(Closeable stream) {
+            throw new CertificateManagerException(ex.getMessage(), ex);
+        } finally {
+            closeStream(is);
+        }
+        return secretStore;
+    }
+
+    public static Map<String, Certificate> getKeystoreMap(final KeyStore keystore) throws KeyStoreException {
+        Map<String, Certificate> retObj = new HashMap<>();
+        Enumeration<String> alaises = keystore.aliases();
+        while(alaises.hasMoreElements()){
+            String alias = alaises.nextElement();
+            Certificate cert = keystore.getCertificate(alias);
+            retObj.put(getCertKeyIdSubject(cert), cert);
+        }
+        return retObj;
+    }
+
+    public static List<Certificate> getChain(Certificate cert, final KeyStore keystore) throws KeyStoreException {
+        return getChain((X509Certificate) cert, getKeystoreMap(keystore));
+    }
+
+    public static List<Certificate> getChain(X509Certificate cert, Map<String, Certificate> keyCache)
+        throws KeyStoreException {
+        List<Certificate> chain = new ArrayList<>();
+        if (null != cert) {
+            chain.add(cert);
+
+            String aki = getCertKeyIdAuthority(cert); // if-root: expected-null
+            while (null != aki) {
+                Certificate certLoop = keyCache.get(aki);
+                chain.add(certLoop);
+                aki = getCertKeyIdAuthority(certLoop);
+            }
+        }
+
+        return chain;
+    }
+
+    public static String getCertKeyIdAuthority(Certificate cert) {
+        return getCertKeyIdAuthority((X509Certificate) cert);
+    }
+    public static String getCertKeyIdAuthority(X509Certificate cert) {
+        String aki = null;
+        if (null != cert) {
+            byte[] akiBytes = cert.getExtensionValue(AUTHORITY_KEY_ID);
+            try {
+                if (akiBytes != null) {
+                    DERDecoder extValA = new DERDecoder(akiBytes);
+                    extValA.skip(6);
+                    aki = Hex.encodeHexString(extValA.getBytes(20));
+                }
+            } catch (WSSecurityException e) {
+                LOG.error("Unable to decode certificate authority {}", e.getLocalizedMessage(), e);
+            }
+        }
+        return aki;
+    }
+
+    public static String getCertKeyIdSubject(Certificate cert) {
+        return getCertKeyIdSubject((X509Certificate) cert);
+    }
+    public static String getCertKeyIdSubject(X509Certificate cert) {
+        String ski = null;
+        if (null != cert) {
+            byte[] subjectKeyID = cert.getExtensionValue(SUBJECT_KEY_ID);
+            try {
+                if (subjectKeyID != null) {
+                    // this logic extracts from CryptoBase class inside wss4j
+                    DERDecoder extVal = new DERDecoder(subjectKeyID);
+                    extVal.expect(DERDecoder.TYPE_OCTET_STRING); // ExtensionValue OCTET STRING
+                    extVal.getLength(); // leave this method alone. getlength modify array position.
```

Comment:
Why you need this call?

Reply:
it required to move the pointer else it return null